src/k8s/pkg/client/kubernetes/upgrades.go

```diff
@@ -0,0 +1,178 @@
+package kubernetes
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"sort"
+
+	"github.com/canonical/k8s/pkg/log"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/runtime/serializer"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/rest"
+)
+
+// TODO: If the upgrade CRD grows, consider using kubebuilder.
+const (
+	UpgradePhaseNodeUpgrade    = "NodeUpgrade"
+	UpgradePhaseFeatureUpgrade = "FeatureUpgrade"
+	UpgradePhaseFailed         = "Failed"
+	UpgradePhaseCompleted      = "Completed"
+)
+
+const (
+	kind    = "Upgrade"
+	group   = "k8sd.io"
+	version = "v1alpha"
+)
+
+type UpgradeMetadata struct {
+	Name string `json:"name,omitempty"`
+}
+
+type UpgradeStatus struct {
+	Phase         string   `json:"phase,omitempty"`
+	UpgradedNodes []string `json:"upgradedNodes,omitempty"`
+}
+
+type Upgrade struct {
+	APIVersion string          `json:"apiVersion,omitempty"`
+	Kind       string          `json:"kind,omitempty"`
+	Metadata   UpgradeMetadata `json:"metadata,omitempty"`
+	Status     UpgradeStatus   `json:"status,omitempty"`
+}
+
+func NewUpgrade(name string) Upgrade {
+	return Upgrade{
+		APIVersion: group + "/" + version,
```

Comment:
nit: Can we use `Sprintf` and why not have this as a constant if we're already joining 2 `const`s?